france_connect/clients.py

```diff
@@ -0,0 +1,192 @@
+            "client_secret": self.client_secret,
+        }
+        try:
+            response = requests.post(
+                token_url,
+                data=data,
+                headers={"Accept": "application/json"},
+                timeout=self.timeout,
+                verify=self.verify_ssl,
+                allow_redirects=self.allow_redirects,
+            )
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.RequestException:
+            msg = f"FranceConnect - Failed to retrieve access_token"
+            raise FranceConnectAccessTokenError(response.status_code, response.text, msg)
+
+    def _verify_jwt(self, token: str, token_type=str) -> dict:
+        """Verify the JWT signature and returns decoded token."""
+        jwks_client = PyJWKClient(f"{self.fc_base_url}{self.fc_jwks_url}")
+        try:
+            # Get the signing key from the JWT
+            key = jwks_client.get_signing_key_from_jwt(token)
+            # Decode and verify the JWT with the signing key
+            return jwt.decode(token, key.key, algorithms=["RS256"], audience=self.client_id)
+        except ExpiredSignatureError:
+            raise FranceConnectExpiredTokenError(f"FranceConnect - FranceConnect {token_type} has expired.")
+        except InvalidTokenError:
+            raise FranceConnectInvalidTokenError(f"FranceConnect - FranceConnect invalid {token_type}")
+
+    def verify_access_token(self, access_token: str) -> dict:
+        """Verify the `access_token`."""
+        return self._verify_jwt(access_token, "access_token")
+
+    def get_user_info(self, token: str) -> dict:
+        """Retrieve user information using `token`."""
+        url = urllib.parse.urljoin(self.fc_base_url, self.fc_userinfo_url)
+        headers = {"Authorization": f"Bearer {token}", "Accept": "application/json"}
+        try:
+            response = requests.get(
+                url, headers=headers, timeout=self.timeout, verify=self.verify_ssl, allow_redirects=self.allow_redirects
+            )
+            response.raise_for_status()
+            return response.content.decode("utf-8")
+        except requests.exceptions.RequestException:
+            msg = f"FranceConnect - Failed to retrieve user_info"
+            raise FranceConnectHttpError(response.status_code, response.text, msg)
+
+    def verify_user_info(self, id_token: str) -> bool:
+        """Verify the user info (id_token) signature."""
+        return self._verify_jwt(id_token, "id_token")
+
+    def get_logout_url(self, id_token: str, state: str, user_info: dict) -> str:
+        """Compute the logout URL for FranceConnect."""
+        post_logout_uri_params = urllib.parse.urlencode(
+            {
+                "given_name": user_info.get("given_name"),
+                "family_name": user_info.get("family_name"),
+                "gender": user_info.get("gender"),
+            }
+        )
```

Comment:
Are these key always in `user_info` even if the user did not provide them in `scopes` ?